### PR TITLE
MorphoClaimRewards in QuarkBuilder

### PR DESCRIPTION
**Pull request by kevincheng96**
This PR implements the QuarkBuilder logic for claiming all unclaimed Morpho rewards for a user.

### CODE REVIEW

File: test/Acceptance/Package.swift [hunk 1]
```diff
@@ -17,7 +17,7 @@ let package = Package(
     ],
     dependencies: [
         // Dependencies declare other packages that this package depends on.
-        .package(url: "https://github.com/hayesgm/Eth.swift", branch: "88a9f560e678fd480f85a38aaee0bef2d1f7e403"),
+        .package(url: "https://github.com/kevincheng96/Eth.swift", branch: "26019982c0f49e9f9c743e9219732188c248300d"),
```
Comment (kevincheng96) [inline test/Acceptance/Package.swift:20]: This [commit](https://github.com/kevincheng96/Eth.swift/commit/26019982c0f49e9f9c743e9219732188c248300d) on my forked repo has the Eth.swift fix we need to support 2D arrays. We can update this back to the upstream repo once that commit is merged.

File: test/Acceptance/Tests/AcceptanceTests/AcceptanceTests.swift [hunk 2]
```diff
@@ -6,17 +6,10 @@ import Testing
...
-    cometWithdrawTests +
-    morphoBorrowTests +
-    morphoRepayTests +
-    morphoVaultSupplyTests +
-    morphoVaultWithdrawTests +
-    swapTests
+let allTests: [AcceptanceTest] =
+    transferTests + cometBorrowTests + cometClaimRewardsTests + cometRepayTests + cometSupplyTests
+    + cometWithdrawTests + morphoBorrowTests + morphoClaimRewardsTests + morphoRepayTests
+    + morphoVaultSupplyTests + morphoVaultWithdrawTests + swapTests
```
Comment (cwang25) [inline test/Acceptance/Tests/AcceptanceTests/AcceptanceTests.swift:12]: Formatting changed?

File: src/builder/actions/Actions.sol [hunk 3]
```diff
@@ -1458,62 +1452,77 @@ library Actions {
...
         internal
         pure
-        returns (IQuarkWallet.QuarkOperation memory, Action memory)
+        returns (IQuarkWallet.QuarkOperation[] memory, Action[] memory)
     {
-        Accounts.ChainAccounts memory accounts =
-            Accounts.findChainAccounts(claimRewards.chainId, claimRewards.chainAccountsList);
+        List.DynamicArray memory quarkOperations = List.newList();
+        List.DynamicArray memory actions = List.newList();
+        // Iterate through each chain and construct a QuarkOperation to claim all rewards for each chain
```
Comment (cwang25) [inline src/builder/actions/Actions.sol:1459]: So we will always "force" to claim rewards on all chains every time it's called?
Probably L2s is fine, but not sure if we want to still keep the option to allow user to select which chains to claim?
(like maybe I only have $0.1 on one chain, and $100 on another chain, and I probably want to only claim $100 one and not bother with $0.1 one.)

File: test/Acceptance/Tests/AcceptanceTests/AcceptanceTests.swift [hunk 4]
```diff
@@ -155,31 +152,21 @@ enum Call: CustomStringConvertible, Equatable {
         }
 
         if scriptAddress == getScriptAddress(CometClaimRewards.creationCode) {
-            if let (cometRewards, comets, accounts) = try? CometClaimRewards.claimDecode(input: calldata) {
+            if let (cometRewards, comets, accounts) = try? CometClaimRewards.claimDecode(
+                input: calldata)
+            {
                 return .claimCometRewards(
-                    cometRewards: cometRewards.map {cometRewardAddress in
+                    cometRewards: cometRewards.map { cometRewardAddress in
```
Comment (cwang25) [inline test/Acceptance/Tests/AcceptanceTests/AcceptanceTests.swift:159]: Some formatter change 🤔 ? it looks like several lines have added extra space in `{xxx` > `{ xxx`

File: src/builder/actions/Actions.sol [hunk 5]
```diff
@@ -1458,62 +1452,77 @@ library Actions {
...
-            nonceSecret: accountSecret.nonceSecret,
-            totalPlays: 1
-        });
+            // Construct QuarkOperation
+            IQuarkWallet.QuarkOperation memory quarkOperation = IQuarkWallet.QuarkOperation({
+                nonce: accountSecret.nonceSecret,
+                isReplayable: false,
+                scriptAddress: CodeJarHelper.getCodeAddress(type(MorphoRewardsActions).creationCode),
+                scriptCalldata: scriptCalldata,
+                scriptSources: new bytes[](0),
```
Comment (cwang25) [inline src/builder/actions/Actions.sol:1498]: Just curious in general that we're passing 0 bytes on scriptSources to save gas on L2 because of the amount of bytes data that will pass in here.  

But what should we do to deploy new scripts to CodeJar? 
Do we send additional transaction on our own behalf with "bytes" in scriptSource before user calling it? 🤔

File: src/builder/actions/Actions.sol (hunk 3 above)
Comment (kevincheng96) [reply src/builder/actions/Actions.sol:1459]: To start, yes. We claim on all chains. But I have a TODO to take in a list of chain ids if the app wants to specify which chains to claim on.

File: src/builder/actions/Actions.sol (hunk 5 above)
Comment (kevincheng96) [reply src/builder/actions/Actions.sol:1498]: We have a deploy script we should run whenever we launch a new chain or support new scripts. Here's the [instructions](https://github.com/legend-hq/legend-scripts?tab=readme-ov-file#deploy)

File: test/Acceptance/Package.swift (hunk 1 above)
Comment (kevincheng96) [reply test/Acceptance/Package.swift]: Updated this back to the `hayesgm/Eth.swift` repo now that a [new release](https://github.com/hayesgm/Eth.swift/releases/tag/0.0.9) is pushed.

File: test/Acceptance/Tests/AcceptanceTests/AcceptanceTests.swift (hunk 2 above)
Comment (kevincheng96) [reply test/Acceptance/Tests/AcceptanceTests/AcceptanceTests.swift]: Un-did all the formatting changes

File: test/Acceptance/Tests/AcceptanceTests/AcceptanceTests.swift (hunk 4 above)
Comment (kevincheng96) [reply test/Acceptance/Tests/AcceptanceTests/AcceptanceTests.swift]: Un-did all the formatting changes

File: src/builder/actions/CometActionsBuilder.sol [hunk 6]
```diff
@@ -192,6 +192,8 @@ contract CometActionsBuilder is QuarkBuilderBase {
         uint256[] memory amountOuts = new uint256[](0);
         string[] memory assetSymbolOuts = new string[](0);
 
+        // TODO: IF WE TAKE IN A LIST OF CHAIN_IDS, we can iterate over the, and then call constructOperationsAndActions on each chain.
+        //       Though to do that, we need to separate out the quote/paycall constructing logic from the action constructing logic
```
Comment (fluffywaffles) [inline src/builder/actions/CometActionsBuilder.sol:196]: SOME YELLING and then we stop :laughing: 
also a dangling particle on the "the," there. Maybe that was meant to be "them"?

File: test/Acceptance/Tests/AcceptanceTests/AcceptanceTests.swift [hunk 7]
```diff
@@ -159,18 +163,6 @@ enum Call: CustomStringConvertible, Equatable {
...
-            {
-                print("supplyTo(\(comet) to: \(to) \(asset) \(amount))")
-            } else if let (comet, from, to, asset, amount) =
-                try? CometSupplyActions.supplyFromDecode(input: calldata)
-            {
-                print("supplyFrom(\(comet) from: \(from) to: \(to) \(asset) \(amount))")
-            } else if let (comet, assets, amounts) =
-                try? CometSupplyActions.supplyMultipleAssetsDecode(input: calldata)
-            {
-                print("supplyMultipleAssets(\(comet) \(assets) \(amounts))")
```
Comment (fluffywaffles) [inline test/Acceptance/Tests/AcceptanceTests/AcceptanceTests.swift:173]: intentional? are these deprecated or deleted functions?

File: test/Acceptance/Tests/AcceptanceTests/AcceptanceTests.swift [hunk 8]
```diff
@@ -1689,6 +1816,30 @@ class Context {
             )
         }
     }
+
+    func reifyMorphoRewardDistributions(network: Network) -> [QuarkBuilder.Accounts
+        .MorphoRewardDistribution]
```
Comment (fluffywaffles) [inline test/Acceptance/Tests/AcceptanceTests/AcceptanceTests.swift:1821]: weird newline?

File: test/Acceptance/Tests/AcceptanceTests/MorphoClaimRewardsTests.swift [hunk 9]
```diff
@@ -0,0 +1,86 @@
+@preconcurrency import Eth
+
+let morphoClaimRewardsTests: [AcceptanceTest] = [
+    .init(
+        name: "Alice claims USDC and WETH rewards from Morpho, paying with QuotePay (testMorphoClaimRewards + testMorphoClaimRewardsWithQuotePay)",
```
Comment (fluffywaffles) [inline test/Acceptance/Tests/AcceptanceTests/MorphoClaimRewardsTests.swift:5]: should we add coverage for some edge cases here? for example:
- invalid proof(s)
- no rewards available
- some distributions invalid vs. all invalid vs. none invalid

**Review by fluffywaffles — APPROVED** [review]
looks good and don't want to block on it, but would like to see some more test coverage on failure cases.

File: test/Acceptance/Tests/AcceptanceTests/MorphoClaimRewardsTests.swift (hunk 9 above)
Comment (kevincheng96) [reply test/Acceptance/Tests/AcceptanceTests/MorphoClaimRewardsTests.swift:5]: These acceptance tests are only testing the output of the `QuarkBuilder`, so there's no way to test invalid proofs. The `QuarkBuilder` just takes in inputs (e.g. `bytes32[][] proofs`) and constructs calldata using these proofs to call the Morpho Legend scripts. The tests for the scripts would be where we would test these edge-cases.
